### ISSUE TITLE
YOLOv10 NMS not working on RK3588

### ISSUE THREAD

**Issue by WhyWouldYouTilt** [ultralytics/ultralytics#18129]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Hey,

im currently facing an issue in using YOLOv10 on an edge device (rokchip rk3588).
When i use the yolov10s.onnx model from https://github.com/airockchip/rknn_model_zoo and follow the example the prediction on the bus.jpg image seems fine and gives me the following results:

![image](https://github.com/user-attachments/assets/7cb87736-f7a5-4d90-a328-2e60dd8a9df6)

The detections look as follows:
IMG: bus.jpg
bus  @ (91 136 555 436) 0.960664
person @ (109 235 226 536) 0.918809
person @ (212 240 284 509) 0.896656
person @ (476 233 559 521) 0.871542
person @ (80 329 124 515) 0.464875

Since i want to train my own model and deploy it on the rokchip i tried in downloading the latest yolov10s ultralytics model in .pt format:
To use this with the rknn model zoo code i did a modification in head.py to fit the model shapes for the rknn model zoo code:

![image](https://github.com/user-attachments/assets/74470515-3159-4da6-aff0-ea96c0bb4edc)

After this conversion my model shapes look exactly the same and i am able to execute both .onnx models on my rokchip.
But when i run the prediction on my yolov10_s_ultralytics.onnx model (same if i convert it to .rknn afterwards) i get the following results for the bus.jpg image:

![image](https://github.com/user-attachments/assets/130d22c9-74a0-466b-845f-7d04749e8dbf)

The prediction look as follows:
IMG: bus.jpg
bus  @ (93 135 555 436) 0.941894
bus  @ (93 135 555 436) 0.932824
bus  @ (91 136 556 437) 0.927678
bus  @ (92 135 556 437) 0.923732
bus  @ (91 136 554 435) 0.920740
bus  @ (92 136 556 436) 0.915927
person @ (109 235 227 535) 0.909801
bus  @ (92 135 555 435) 0.908342
person @ (109 235 226 535) 0.907098
person @ (110 235 226 535) 0.896609
bus  @ (91 135 551 435) 0.893832
person @ (110 235 226 535) 0.890245
person @ (110 236 226 535) 0.882031
person @ (109 235 226 535) 0.881972
person @ (109 235 227 535) 0.878468
person @ (110 236 226 537) 0.876065
person @ (476 235 559 522) 0.873481
person @ (109 236 226 536) 0.867698
person @ (109 235 226 535) 0.866305
person @ (476 233 559 522) 0.863528
person @ (476 232 559 521) 0.862647
person @ (476 233 559 522) 0.860216
person @ (475 234 559 522) 0.858624
person @ (476 233 559 522) 0.858264
bus  @ (90 135 553 436) 0.858161
person @ (476 235 559 522) 0.852844
person @ (476 233 559 522) 0.849531
person @ (211 240 285 509) 0.844894
bus  @ (92 135 554 435) 0.843959
person @ (211 240 285 508) 0.840483
person @ (212 241 284 510) 0.839818
person @ (211 241 285 510) 0.838602
person @ (211 240 284 509) 0.826926
person @ (476 232 559 522) 0.825879
person @ (211 239 284 510) 0.820333
person @ (211 240 284 508) 0.814625
person @ (210 241 285 511) 0.801479
person @ (212 240 284 509) 0.798915
person @ (212 240 284 509) 0.797419
person @ (476 236 559 521) 0.688462
person @ (80 327 123 515) 0.508772
person @ (80 327 123 514) 0.498710
person @ (80 330 123 515) 0.471755
person @ (79 327 124 515) 0.468778
person @ (80 327 123 514) 0.465623
person @ (80 331 125 515) 0.451433
person @ (80 327 122 513) 0.427664
person @ (80 326 123 514) 0.414136
person @ (210 240 284 509) 0.339251
person @ (80 328 121 512) 0.315434

I was thinking that for some reason no NMS is applied, but since yolov10 is NMS free i am kind of confused... The found predictions are refering to the same objects and they nearly have the same score.

It would be really nice if someone could help me here since i am stuck at this problem.

Best regards
Marvin

### Environment

Ultralytics 8.3.28 🚀 Python-3.11.4 torch-2.5.1+cpu CPU (unknown)

### Minimal Reproducible Example

    def forward(self, x):
        if torch.onnx.is_in_onnx_export():
            return self.forward_export(x)
        if self.end2end:
            return self.forward_end2end(x)

        for i in range(self.nl):
            x[i] = torch.cat((self.cv2[i](x[i]), self.cv3[i](x[i])), 1)
        if self.training:  # Training path
            return x
        y = self._inference(x)
        return y if self.export else (y, x)
    
    def forward_export(self, x):
        results = []
        for i in range(self.nl):
            dfl = self.cv2[i](x[i]).contiguous()
            cls = self.cv3[i](x[i]).contiguous()
            cls = torch.sigmoid(cls)
            results.append(dfl)
            results.append(cls)
        return tuple(results)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @WhyWouldYouTilt, thank you for bringing this up and for your detailed report! 🚀 

We appreciate your interest in using Ultralytics YOLOv10 on edge devices like the RK3588. To assist you better, we’d ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) (MRE). This helps us debug your issue more effectively and identify potential root causes.

A few suggestions and checks in the meantime:
1. Ensure you are using the latest version of the `ultralytics` package, as improvements or fixes might already address the issue. You can upgrade with:
    ```bash
    pip install -U ultralytics
    ```
2. Please double-check your modifications in `head.py` as even small discrepancies in layer configurations or model architecture might introduce unexpected behaviors. 
3. Verify that your exported `ONNX` model aligns with [Ultralytics export guidelines](https://docs.ultralytics.com/modes/export/) for inference.

If this is related to a 🔧 custom integration or deployment issue on RK3588, additional details such as logs, environment configurations, or snippets of the **conversion pipeline** would be helpful to debug your specific case. If NMS behavior is suspected, verifying the intermediate model outputs pre- and post-conversion can shed light on potential differences.

For further discussions, you can also engage with the Ultralytics community for insights from others working on similar setups:
- Join our [Discord](https://discord.com/invite/ultralytics) to discuss this in real time 🎧
- Post on our [Discourse forums](https://community.ultralytics.com) for an in-depth conversation
- Explore our [Reddit](https://reddit.com/r/Ultralytics) for sharing experiences and solutions with the community

## Environments

For debugging, we recommend testing in these verified YOLO environments to rule out external factors:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Additionally:
- If you're running your project in a constrained environment like the RK3588, ensure all dependencies (especially [PyTorch](https://pytorch.org/) and export-related libraries like `onnxruntime`) are at compatible versions.

---

This is an automated response to help streamline your experience 😊. An Ultralytics engineer will review your issue shortly and provide additional insights. Thank you for your patience and for being part of the Ultralytics community! 🚀

**Comment by Y-T-G** [comment]
This is the forward function that needs to be run for YOLOv10.

https://github.com/ultralytics/ultralytics/blob/420fba7fb2fab2bed8d6cea6a8d0095ce3f9c123/ultralytics/nn/modules/head.py#L76

You're skipping it.

Your model is returning results from one2many head which behaves the same as YOLOv8 head. You need the results from one2one head.

**Comment by WhyWouldYouTilt** [comment]
Thank you for your response, but when i run that function for the export i get the following model shape:
![image](https://github.com/user-attachments/assets/6067248d-7ad0-48f5-8b51-f3bf2615c1e3)
When i export this model to .rknn and execute it with rknntoolkitlite2 on my RK3588 i run into a segmentation fault:
![image](https://github.com/user-attachments/assets/8655e60b-9826-4964-8edd-611fbcbf4afa)

which is probably caused by the missmatch of the shapes. Thats why i changed the head.py as i stated in my previous comment. The rknn_model_zoo models have the following shape compared to the ultralytics export with using forward_end2end:
![image](https://github.com/user-attachments/assets/c1251a00-c9e6-4f5d-ad39-2a9e41a29600)

**Comment by Y-T-G** [comment]
You can try replacing `self.cv2` with `self.one2one_cv2` and `self.cv3` with `self.one2one_cv3` in your modified forward method.

**Comment by WhyWouldYouTilt** [comment]
Thank you very much, it worked with that change! :) Can you shortly explain why this lead to the fix? It would really help me understanding the model more...

**Comment by Y-T-G** [comment]
It's what I mentioned before:
> Your model is returning results from one2many head which behaves the same as YOLOv8 head. You need the results from one2one head.

**Comment by WhyWouldYouTilt** [comment]
True you are right, i also took another look at the yolov10 paper where this is also explained. I will close the issue since it is solved. Thank you :)

**Comment by glenn-jocher** [comment]
You're welcome! Glad the issue is resolved. If you have any further questions, feel free to reach out to the YOLO community or check the [YOLOv10 documentation](https://docs.ultralytics.com/models/yolov10/).
